common/workunit/workunit.cpp

```diff
@@ -5462,9 +5490,70 @@ class CDaliWorkUnitFactory : public CWorkUnitFactory, implements IDaliClientShut
+
+                    bool dynamicAbortingState = false;
+                    const char* wuid = cur.queryName();
+                    WUState state = getState(&cur, wuAbortsPTree, dynamicAbortingState);
+                    if (!dynamicAbortingState)
+                        lWWUPTreeRoot->addPropTree(wuid, LINK(&cur));
+                    else
+                    {
+                        Owned<IPropertyTree> lWWUPTree = createLightWeitghtWorkunitPTree(&cur, state);
+                        lWWUPTreeRoot->addPropTree(wuid, lWWUPTree.getClear());
+                        hasDynamicAbortingState = true;
+                    }
+                }
+                return hasDynamicAbortingState ? lWWUPTreeRoot->getElements("*") : nullptr;
+            }
+            IPropertyTree* createLightWeitghtWorkunitPTree(IPropertyTree* originalWUTree, WUState& state)
+            {
+                Owned<IPropertyTree> lWWUPTree = createPTree();
+                lWWUPTree->setProp("@jobName", originalWUTree->queryProp("@jobName"));
+                lWWUPTree->setProp("@submitID", originalWUTree->queryProp("@submitID"));
+                lWWUPTree->setProp("@clusterName", originalWUTree->queryProp("@clusterName"));
+                lWWUPTree->setProp("@timeScheduled", originalWUTree->queryProp("@timeScheduled"));
+                lWWUPTree->setProp("@scope", originalWUTree->queryProp("@scope"));
+                lWWUPTree->setProp("Action", originalWUTree->queryProp("Action"));
+                lWWUPTree->setProp("@totalThorTime", originalWUTree->queryProp("@totalThorTime"));
+                lWWUPTree->setProp("@priorityClass", originalWUTree->queryProp("@priorityClass"));
+                lWWUPTree->setPropBool("@protected", originalWUTree->getPropBool("@protected"));
+                setEnum(lWWUPTree, "State", state, states);
+
+                Owned<IPropertyTree> application = originalWUTree->getBranch("Application");
+                if (application)
+                    lWWUPTree->addPropTree("Application", application.getClear());
+                return lWWUPTree.getClear();
+            }
+            WUState getState(IPropertyTree* originalWUTree, IPropertyTree* wuAbortsPTree, bool& dynamicAbortingState)
+            {
+                //In the CLocalWorkUnit::getState(), the checkAgentRunning() is used to get the latest
+                //state for one WU using Dali session manager. But, this getState() may give an
+                //out-of-date state in some unusual circumstances because no effective way can be used
+                //to get the latest state using Dali session manager.
+                const char* wuid = originalWUTree->queryName();
+                auto aborting = [wuid, wuAbortsPTree] ()
+                {
+                    return wuAbortsPTree->queryPropTree(wuid) ? true : false;
+                };
+                return getWUState(originalWUTree, aborting, dynamicAbortingState);
```

Comment:
minor: as per above comments, as this method is really only defining a 1 line lambda function, I'd move to inline above.

Reply:
ok